### PR TITLE
Use pytest instead of unittest

### PR DESCRIPTION
**Pull request by Batalex**
See https://github.com/canonical/operator/issues/1110

Libs needed an upgrade, I am not sure how we should proceed in this case

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -7,51 +7,61 @@
...
-                    "startup": "enabled",
-                    "environment": {
-                        "DEMO_SERVER_DB_HOST": None,
-                        "DEMO_SERVER_DB_PASSWORD": None,
-                        "DEMO_SERVER_DB_PORT": None,
-                        "DEMO_SERVER_DB_USER": None,
-                    },
-                }
+
+@pytest.fixture(autouse=True)
```
Comment (benhoyt) [inline tests/unit/test_charm.py]: I'm not a big fan of autouse=True fixtures, as they seem too magical (automatically applied decorators, essentially). It's sort of okay here because they're quite close to the code that's using them, but I think we should avoid them, especially in examples. So let's just go back to plain old `@patch(...)` for these two.

**Review by benhoyt — CHANGES_REQUESTED** [review]
Great, thanks -- much easier to read. Two things:

1) Let's not try to upgrade the libs (at least not in the same PR). I think we can just remove those changes from this.
2) This repo has many branches for all the different parts of the tutorial. Does this need to be updated on multiple branches? (We know this is a bit annoying -- Tony and I are in discussion with the documentation folks to see if we can improve this.)

File: tests/unit/test_charm.py (hunk 1 above)
Comment (tonyandrewmeyer) [reply tests/unit/test_charm.py]: I'm not a big fan of `autouse=True` either, but I have an alternative suggestion: just use the `monkeypatch` fixture in the tests, like:

```python
def test_pebble_layer(monkeypatch, harness):
    monkeypatch.setattr(FastAPIDemoCharm, "version", "1.0.0")
    ...
```

I feel like `monkeypatch` is more "pytest" than `unittest.mock`.

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -7,51 +7,61 @@
...
+    # Simulate the container coming up and emission of pebble-ready event
+    harness.container_pebble_ready("demo-server")
+    # Get the plan now we've run PebbleReady
+    updated_plan = harness.get_container_pebble_plan("demo-server").to_dict()
+    service = harness.model.unit.get_container("demo-server").get_service(
+        "fastapi-service"
+    )
+    status = harness.model.unit.status
+
+    # Check that we have the plan we expected
```
Comment (tonyandrewmeyer) [inline tests/unit/test_charm.py]: I know this was there in the original, but taking the opportunity for a drive-by:

```suggestion
    # Check that we have the plan we expected:
```

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -7,51 +7,61 @@
...
+    assert updated_plan == expected_plan
+    # Check the service was started
```
Comment (tonyandrewmeyer) [inline tests/unit/test_charm.py]: And here:

```suggestion
    # Check the service was started:
```

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -7,51 +7,61 @@
...
+    assert service.is_running()
+    # Ensure we set an ActiveStatus with no message
```
Comment (tonyandrewmeyer) [inline tests/unit/test_charm.py]: And here:

```suggestion
    # Ensure we set an ActiveStatus with no message:
```

File: tests/unit/test_charm.py [hunk 5]
```diff
@@ -7,51 +7,61 @@
...
+    }
+
+    # Simulate the container coming up and emission of pebble-ready event
+    harness.container_pebble_ready("demo-server")
+    # Get the plan now we've run PebbleReady
+    updated_plan = harness.get_container_pebble_plan("demo-server").to_dict()
+    service = harness.model.unit.get_container("demo-server").get_service(
+        "fastapi-service"
+    )
+    status = harness.model.unit.status
```
Comment (tonyandrewmeyer) [inline tests/unit/test_charm.py]: I'm ok with the reorganising from "get the thing, check the thing" to "get all the things, check all the things", but I'm not that convinced that this one line adds much over just having `assert harness.model.unit.status == ops.ActiveStatus()` - but if you feel strongly that it's better then that's ok.

Comment (Batalex) [reply tests/unit/test_charm.py]: Inlined in 56c7028!

File: tests/unit/test_charm.py (hunk 1 above)
Comment (Batalex) [reply tests/unit/test_charm.py]: Replaced auto fixtures with explicit calls in 31b645a.

**Comment by Batalex** [comment]
I reverted the changes to libs.  We could add those changes in `04_integrate_with_psql` and cascade-rebase subsequent branches maybe? (and update `ops` lower constraint as well)

**Review by benhoyt — APPROVED** [review]
Looks good to me now, thanks for the updates!

**Review by tonyandrewmeyer — APPROVED** [review]
